Solve ode (ordinary differential equation) guide page

bertiewooster opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->


#### Brief description of what is fixed or changed
Add guide page

#### Other comments


#### Release Notes

<!-- Write the release notes for this release below between the BEGIN and END
statements. The basic format is a bulleted list with the name of the subpackage
and the release note for this PR. For example:

* solvers
  * Added a new solver for logarithmic equations.

* functions
  * Fixed a bug with log of integers.

or if no release note(s) should be included use:

NO ENTRY

See https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more
information on how to write release notes. The bot will check your release
notes automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->
* other
  * A new, user-friendly page on solving a system of equations algebraically was added to the Guides section.
<!-- END RELEASE NOTES -->

bertiewooster committed:
Initial content for ode

bertiewooster committed:
Add note about which SciPy ODE solver functions to link to

bertiewooster committed:
Merge branch 'master' into solve-ode

bertiewooster commented:
This is a very early draft. I have little experience in ODEs, so I'd appreciate guidance on:

- Which [SciPy](https://docs.scipy.org/doc/scipy/reference/integrate.html?highlight=ode) (or other free and open-source package) function(s) are alternatives, for example for numerical solutions
- Which [`dsolve`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#sympy.solvers.ode.dsolve) parameters to discuss
    - from StackOverflow, these may be worth mentioning:
        - `ics` initial conditions
        - `hint` solving method that you want dsolve to use
    - Unsure if these are worth discussing:
        - `simplify` using `odesimp`
        - `xi` and `eta` infinitesimal functions
        - `x0` point about which the power series solution of a differential equation is to be evaluated
        - `n` exponent of the dependent variable up to which the power series solution of a differential equation is to be evaluated
- Which [other ODE functions](https://docs.sympy.org/dev/modules/solvers/ode.html) to include
    - [`checkodesol`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#checkodesol) -- I already included in the example to check that SymPy's solution is correct
    - [`classify_ode`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#classify-ode) is [mentioned on StackOverflow](https://stackoverflow.com/questions/57678835/is-sympy-correctly-solving-this-ode) in the context of `hint`
    - [`homogeneous_order`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#homogeneous-order) Returns the order if is homogeneous and None if it is not homogeneous.
    - [`infinitesimals`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#infinitesimals) -- this seems pretty technical
    - [`checkinfsol`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#checkinfsol) -- likewise
    - [`constantsimp`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#constantsimp) Simplifies an expression with arbitrary constants in it.
 - Can, and should, you use [`dsolve`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#sympy.solvers.ode.dsolve) solve a system of ODEs? The existing documentation indicates yes, but then describes [`dsolve_system`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#dsolve-system) as Solves any(supported) system of Ordinary Differential Equations.

bertiewooster commented:
Was it the intention to make `dsolve_system` a user function as listed on the [ODE page](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#dsolve-system)? [`dsolve_system` is called by `dsolve` for iterables](https://github.com/sympy/sympy/blob/master/sympy/solvers/ode/ode.py#L559) (multiple equations); is there a reason for users to call `dsolve_system` instead of `dsolve`?

bertiewooster committed:
Convert example from f(x) to y(x)

bertiewooster commented:
@asmeurer When we met, I recall you discussing using `Derivative` vs. `diff` when expressing derivatives in differential equations to be solved, but my notes don't include what the recommendation is. Both seem to work, for example

```
>>> from sympy import Function, dsolve, diff
>>> result = dsolve(diff(y(x), x, x) + 9*y(x), y(x)); result
Eq(y(x), C1*sin(3*x) + C2*cos(3*x))
```

gives the same result as using `Derivative` instead of `diff`. And the existing [`dsolve` docstring](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#dsolve) uses some of each.

Is there a recommendation about using `Derivative` vs. `diff`?

oscarbenjamin commented:
The difference between Derivative and diff is whether SymPy actually tries to compute the derivative or just give an unevaluated derivative:
```python
In [5]: Derivative(x**2, x)
Out[5]: 
d ⎛ 2⎞
──⎝x ⎠
dx    

In [6]: diff(x**2, x)
Out[6]: 2⋅x

In [7]: (x**2).diff(x)
Out[7]: 2⋅x
```
For an undefined function like `f(x)` it isn't possible to evaluate the derivative so it doesn't make any difference.

It's worth noting that a lot of users don't seem to like writing `f(x)` explicitly and prefer to do something like:
```python
In [8]: t = symbols('t')

In [9]: x = Function('x')(t)

In [10]: x
Out[10]: x(t)

In [11]: xp = x.diff()

In [12]: xpp = xp.diff()

In [13]: eq = xpp + 2*xp + x

In [14]: eq
Out[14]: 
                      2      
         d           d       
x(t) + 2⋅──(x(t)) + ───(x(t))
         dt           2      
                    dt       

In [15]: dsolve(eq)
Out[15]: 
                    -t
x(t) = (C₁ + C₂⋅t)⋅ℯ  
```
It would be nice to have a helper function for this like:
```python
x, xp, xpp = function_derivatives('x', t, 2)
```

bertiewooster commented:
### `Derivative` vs. `diff`
I wondered whether we might want to recommend `Derivative` in case the user has already defined the function, so I checked whether using `Derivative` vs. `diff` makes any difference in that case, and it seems like not--both give the same error:

```py
>>> from sympy import Function, dsolve, Derivative, diff
>>> from sympy.abc import x
>>> y = Function('y')(x)
>>> y = x**2; y
x**2
>>> result = dsolve(Derivative(y(x), x, x) + 9*y(x), y(x)); result
TypeError: 'Pow' object is not callable
>>> result = dsolve(diff(y(x), x, x) + 9*y(x), y(x)); result
TypeError: 'Pow' object is not callable
```

If there's no reason to recommend `Derivative` vs. `diff`, maybe I'll just write to use `Derivative` to express the unevaluated derivative, and not recommend against using `diff`?

### Ways to express `f(x)`
> It's worth noting that a lot of users don't seem to like writing `f(x)` explicitly and prefer to do something like:
> ...
> `x = Function('x')(t)`
>...

Is that a positive (or at least decent) pattern we should include in this guide, or an anti-pattern that we want to discourage and thus should not include?

oscarbenjamin commented:
> Is that a positive (or at least decent) pattern we should include in this guide, or an anti-pattern that we want to discourage and thus should not include?

I don't think there's any reason to encourage or discourage it but the two different ways of doing it subtly affect the code needed e.g. for specifying the the initial conditions. It also causes problems for printing and reevaluating expressions because the `x` variable prints as `x(t)` but calling `x(t)` in the code would be an error so any printed expression cannot be used to copy and paste. This is mostly just a matter of taste but it is something that leads to beginner confusion so maybe it's worth showing how to do it both ways at least for one example.

bertiewooster commented:
Do folks think it's worth discussing `hint` and `classifiy_ode` (they're related topics, about how to try to solve an ODE)? @asmeurer was unsure, so I'd appreciate others' thoughts.

oscarbenjamin commented:
> Do folks think it's worth discussing `hint` and `classifiy_ode`

I think it is but I would leave it to be the last thing that is discussed.

bertiewooster commented:
> Was it the intention to make `dsolve_system` a user function as listed on the [ODE page](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#dsolve-system)? [`dsolve_system` is called by `dsolve` for iterables](https://github.com/sympy/sympy/blob/master/sympy/solvers/ode/ode.py#L559) (multiple equations); is there a reason for users to call `dsolve_system` instead of `dsolve`?

I compared the docstrings and `dsolve_system` does two parameters that `dsolve` does not:

> t : Symbol or None
> Independent variable in the system of ODEs
> 
> doit : Boolean
> Evaluate the solutions if True. Default value is True. Can be set to false if the integral evaluation takes too much time and/or is not required.

so it seems like there's a case for having `dsolve_system` as a user function. (Someone knowledgeable about those two functions might be able to judge whether those two parameters could and should be added to `dsolve` so that there would be no need to have `dsolve_system` as a user function.)

Considering that `dsolve_system` has just those two parameters, I'm inclined to

- include `dsolve_system` on the same guide page as `dsolve`
- include `dsolve_system` only in examples using those two parameters.

Let me know if you think otherwise.

bertiewooster commented:
> ics ... should be given in the form of `{f(x0): x1, f(x).diff(x).subs(x, x2): x3}`

For the first dictionary entry, I understand that it's setting the value of f at x-value x0 (e.g. 0) to the value x1 (e.g. 1), e.g. f(0) = 1.

I'm having trouble understanding the second entry, specifically why it needs two constants x2 and x3. For example, substituting x2 = 2 gives a constant, `-sin(2)`:

```py
>>> from sympy import cos
>>> f = Function('f')(x); f
f(x)
>>> f = cos(x); f
cos(x)
>>> f.diff(x)
-sin(x)
>>> f.diff(x).subs(x, 2)
-sin(2)
```

so why does there need to be a second constant x3? I wondered if it's because f could be a function of multiple variables, but that seems to contradict the f(x) notation.

bertiewooster committed:
Rearrange by how define input function
Explain how to extract result from Equality

bertiewooster committed:
Restructure later sections

bertiewooster committed:
Add warning about copying and pasting results
if define a function of an independent variable

bertiewooster committed:
Add section Work With Arbitrary Constants

bertiewooster commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,270 @@
+C1 - C2*exp(x)
+```
+
+### Work With Arbitrary Constants
+
+When you want to simplify arbitrary constants such as `C1`, `C2`, `C3`, etc.,
+use {func}`~.constantsimp`. For example, if you add two ODE solutions together,
+there may be an opportunity to simplify the constants:
+
+```py
+>>> from sympy import Function, dsolve, Derivative, symbols
+>>> from sympy.abc import x
+>>> from sympy.solvers.ode.ode import constantsimp
+>>> y = Function('y')
+>>> result = dsolve(Derivative(y(x), x, x) + 9*y(x), y(x)); result
+Eq(y(x), C1*sin(3*x) + C2*cos(3*x))
+>>> sol = result.rhs; sol
+C1*sin(3*x) + C2*cos(3*x)
+>>> ex = sol + (3 * sol / 2); ex
+5*C1*sin(3*x)/2 + 5*C2*cos(3*x)/2
+>>> C1, C2, C3 = symbols("C1 C2 C3")
+>>> constantsimp(ex, {C1, C2, C3})
+C1*sin(3*x) + C2*cos(3*x)
+```
+
+After simplifying the arbitrary constants, those constants may be
+non-consecutive ordered. For example the simplified result may contain `C1` and
+`C3` but not `C2`. You can fix this using {func}`~.constant_renumber`:
+
+```py
+>>> from sympy import symbols, cos, sin
+>>> from sympy.solvers.ode.ode import constantsimp, constant_renumber
+>>> C1, C2, C3 = symbols("C1 C2 C3")
+>>> expr = C1*cos(x) + C2*cos(x) + C3*sin(x); expr
+C1*cos(x) + C2*cos(x) + C3*sin(x)
+>>> simplified = constantsimp(expr, {C1, C2, C3}); simplified
+C1*cos(x) + C3*sin(x)
+>>> constant_renumber(simplified)
+C1*sin(x) + C2*cos(x)
+```
````
I was trying to come up with realistic examples of when you would work with arbitrary constants. Perhaps it's my ignorance, but I didn't find the examples in [`constantsimp`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=constantsimp#sympy.solvers.ode.constantsimp) or [`constant_renumber`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=constantsimp#sympy.solvers.ode.ode.constant_renumber) very compelling because they don't come from `dsolve` results (they just start by creating expressions involving `C1`, `C2`, `C3`) despite `constantsimp` being "written specifically to work with [dsolve()](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=constantsimp#sympy.solvers.ode.dsolve)". I understand the statement "There is no need to specify the dependent function, such as f(x), because it already has the independent symbol, x, in it." but think the examples would follow more naturally if they involved functions as `dsolve` typically returns such as $\cos(x)$.

For `constantsimp`, the best I came up with was adding one solution to a multiple of itself. Is there a more realistic situation where arbitrary constants could be simplified?

For `constant_renumber`, is there some example where applying `constantsimp` leads to removal of an arbitrary constant?

asmeurer replied:
I don't think we should discuss either of these functions. They can be useful for end users in some situations, like if you are solving an ODE that comes from another ODE, so that you end up with constants from both solutions, but those situations are not typical. 

`constantsimp` and `constant_renumber` are already applied automatically to the result of `dsolve` (unless you specify `simplify=False`; and unlike `solve`, there's rarely a reason to need to do that).

bertiewooster committed:
Use ~ to remove path to subs()

asmeurer commented:
> Is that a positive (or at least decent) pattern we should include in this guide, or an anti-pattern that we want to discourage and thus should not include?

I've never been a huge fan of it, and you won't really see it used in any other SymPy documentation. The main reason is that it's confusing if the variable `f` points to `f(x)` instead of `f` itself, so if you try `f(0)` you'll get an error. A lot of it comes down to how comfortable people are with how variables work in Python (which we have some gotchas guides on). If you understand what is going on, it's probably fine, although you're still prone to forget. 

But as Oscar mentioned, it is used by people a lot, because people just don't like typing out `f(x)` all the time, so it might be worth mentioning. Assigning the derivative to a variable is also useful because f(x).diff(x) is a lot to type. For a guide, it's easy to jump in in the middle and see the variable `f` being used and have it not be clear that `f` is actually defined as `f(x)` instead of plain `Function('f')`. So I would just mention this, but keep the verbose style otherwise.

Also, not sure how obvious it is that these are all equivalent:

```
xpp = f(x).diff(x, x)
xpp = f(x).diff(x, 2)
xpp = f(x).diff((x, 2))
xpp = diff(f(x), x, x)
xpp = diff(f(x), x, 2)
xpp = Derivative(f(x), x, x)
xpp = Derivative(f(x), x, 2)
xpp = Derivative(Derivative(f(f(x), x), x)
xpp = diff(diff(f(x), x), x)
xp = f(x).diff(x)
xpp = xp.diff(x)
```

Regarding `dsolve_system`, we should just recommend `dsolve`. `dsolve_system` is not a publicly imported function, so it should be viewed as an internal helper for `dsolve`, which is what should be used by end-users.

asmeurer commented:
> Do folks think it's worth discussing hint and classifiy_ode

The most important thing here is the `_Integral` hints, which let you get the result of the ODE with unevaluated integrals.

They also let you see/use different solvers that can be applied to the same ODE. This can be useful for primarily two reasons:

1. Educational purposes, like if you are learning about a specific method to solve ODEs and want to get a result that exactly matches that method.
2. Because sometimes an ODE can be solved by many different solvers, and they can return different results. They will be mathematically equivalent, though often the constants won't be. dsolve by default tries to use the "best" solvers first, which are most likely to return the most usable output, but it's not a perfect heuristic. Something that can sometimes happen is one solver will produce a result with an Integral that SymPy cannot solve but another solver will produce a different integral which it can. So in general, if the solution isn't in a form you like, you can try other hints to see if they give something better. 

Good examples of this are ODEs that match the homogeneous_order hints.

bertiewooster committed:
Remove discussion of constantsimp and constant_renumber

bertiewooster committed:
Remove dsolve_system. Give many ways to express second derivative.

bertiewooster committed:
Note which hint to write up

bertiewooster committed:
Add section Ordinary Differential Equation Solving Hints

bertiewooster commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,262 @@
+supply: for any hint with a corresponding `_Integral` hint, `all_Integral` only
+returns the `_Integral` hint.
+
+### Select a Specific Solver
+
+You may wish to select a specific solver using a hint for a couple of reasons:
+- educational purposes: for example if you are learning about a specific method
+  to solve ODEs and want to get a result that exactly matches that method
+- form of the result: sometimes an ODE can be solved by many different solvers,
+  and they can return different results. They will be mathematically equivalent,
+  though the arbitrary constants may not be. {func}`~.dsolve` by default tries
+  to use the "best" solvers first, which are most likely to return the most
+  usable output, but it is not a perfect heuristic. For example, the "best"
+  solver may produce a result with an integral that SymPy cannot solve, but
+  another solver may produce a different integral that SymPy can solve. So if
+  the solution isn't in a form you like, you can try other hints to check
+  whether they give a preferable result.
```
> Good examples of this [using different solvers that can be applied to the same ODE] are ODEs that match the homogeneous_order hints

@asmeurer is it worth giving an example of this? If so, is one of the homogeneous order equations given in the [`homogeneous_order`](https://docs.sympy.org/dev/modules/solvers/ode.html?highlight=ode#homogeneous-order) docstring a good example, either `f(x)` or `x**2*f(x)/sqrt(x**2+f(x)**2)`?

bertiewooster replied:
@asmeurer, reminding you to please construct an example where the form of the ODE solution depends on the hint given.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,262 @@
+
+Alternatives to consider:
+- *which SciPy functions? from
+  https://docs.scipy.org/doc/scipy/reference/integrate.html?highlight=ode?*
```
I think solve_ivp is good for most instances https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp. Or we can just reference this page https://docs.scipy.org/doc/scipy/reference/integrate.html#solving-initial-value-problems-for-ode-systems

bertiewooster committed:
Work With Arbitrary Constants

bertiewooster committed:
Equations With No Solution

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,262 @@
+
+*Equations with no solution content*
+
+### Equations With No Analytical Solution
```
This looks like a bug, although for a system you should specify the list of functions as the second argument to dsolve (also a good idea for single ODEs, although it will do it automatically if there is only one function in the expression).

oscarbenjamin replied:
That's one of the legacy solvers that are mostly just broken and should probably just be removed:
https://github.com/sympy/sympy/wiki/ODE-Systems-roadmap#nonlinear-solvers---eliminating-the-independent-variable

bertiewooster replied:
Thanks; I added that you supply a list of functions to solve for.

I'm still seeking an example of an ODE (or system of) where there isn't a closed-form solution (and SymPy recognizes that), and would appreciate any ideas!

asmeurer replied:
Maybe a double pendulum. See https://scienceworld.wolfram.com/physics/DoublePendulum.html (eqs 14 and 19). Here I simplified them by setting the masses and lengths to 1, although I'm sure it would give the same thing too if I included them:

```py
>>> theta1, theta2 = symbols('theta1 theta2', cls=Function)
>>> g = symbols('g')
>>> eq1 = 2*theta1(t).diff(t, t) + theta2(t).diff(t, t)*cos(theta1(t) - theta2(t)) + theta2(t).diff(t)**2*sin(theta1(t) - theta2(t)) + 2*g*sin(theta1(t))
>>> eq2 = theta2(t).diff(t, t) + theta1(t).diff(t, t)*cos(theta1(t) - theta2(t)) - theta1(t).diff(t)**2*sin(theta1(t) - theta2(t)) + g*sin(theta2(t))
>>> dsolve([eq1, eq2], [theta1(t), theta2(t)])
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "/Users/aaronmeurer/Documents/Python/sympy/sympy/./sympy/solvers/ode/ode.py", line 589, in dsolve
    raise NotImplementedError
NotImplementedError
```

(please double check that I transcribed it correctly)

bertiewooster committed:
Demonstrate how to supply multiple functions to be solved for
Include function(s) to be solved for in all examples

bertiewooster commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,262 @@
+
+### Equations Which Have An Analytical Solution, and SymPy Cannot Solve
+
+*Equations which have an analytical solution, and SymPy cannot solve content*
```
Can someone (perhaps @oscarbenjamin) come up with an ODE (one or system) which has a closed-form solution, but which SymPy cannot solve?

oscarbenjamin replied:
Any example like that will always be fragile because in principle a future version of SymPy will be able to solve it. In fact putting it in the guide makes it particularly likely that `dsolve` would be improved in future to handle that case.

bertiewooster committed:
Must specify argument of function to solve for

bertiewooster committed:
Give example of setting initial condition of derivative

bertiewooster committed:
Make Alternatives to consider a heading, link to it as temporary workaround

bertiewooster committed:
Fill in links to SciPy numerical ODE solvers

bertiewooster committed:
Trying to get Sphinx on CircleCI to recognize scipy.integrate.solve_ivp,
which works locally

bertiewooster committed:
Remove SciPy intersphinx link because not working,
use URL instead

bertiewooster committed:
Mention Lorenz system as example of no closed-form solution

bertiewooster committed:
Add double pendulum example, If You Encounter a Problem With SymPy ODE Solving

bertiewooster committed:
Write out double pendulum equations in math format

moorepants commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,341 @@
+## Alternatives to Consider
+- To numerically solve a system of ODEs, use a [SciPy ODE
+  solver](https://docs.scipy.org/doc/scipy/reference/integrate.html#solving-initial-value-problems-for-ode-systems)
+  such as `solve_ivp`.
```
A super common thing to do is to create an ode in sympy, lambdify it for use in solve_ivp, and then integrate numerically. I think it could be helpful for us to show that here (or another guide). We have some examples in the code generation tutorial: https://www.sympy.org/scipy-2017-codegen-tutorial/ and here is an example from my course materials: https://moorepants.github.io/learn-multibody-dynamics/simulation.html.

bertiewooster replied:
Thanks, @moorepants. I would be happy to include such an example on this page. I'm having a tough time converting an [example from the code generation tutorial](https://www.sympy.org/scipy-2017-codegen-tutorial/notebooks/25-chemical-kinetics-intro.html#Solving-the-initial-value-problem-numerically) from [odeint](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.odeint.html?highlight=odeint#scipy.integrate.odeint) (SciPy's outdated function) to [solve_ivp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp) and using [lambdify](https://docs.sympy.org/latest/modules/utilities/lambdify.html?highlight=lambdify#module-sympy.utilities.lambdify). Here's what I've tried and the error it gives. Any assistance would be greatly appreciated.

```py
import sympy as sym
sym.init_printing()

def rhs(y, t, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, k = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(y, None, *k)

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((y, t) + k, ydot)  # EXERCISE: (y, t) + k

import numpy as np
from scipy.integrate import solve_ivp
tout = np.linspace(0, 10)
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, tout, y0, args=k_vals)
```

```py
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
/var/folders/9w/cn8tqbv976gckxd0c9cbr1sw0000gn/T/ipykernel_96785/228934301.py in <module>
----> 1 solve_ivp(f, tout, y0, args=k_vals)

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/ivp.py in solve_ivp(fun, t_span, y0, method, t_eval, dense_output, events, vectorized, args, **options)
    540         method = METHODS[method]
    541 
--> 542     solver = method(fun, t0, y0, tf, vectorized=vectorized, **options)
    543 
    544     if t_eval is None:

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/rk.py in __init__(self, fun, t0, y0, t_bound, max_step, rtol, atol, vectorized, first_step, **extraneous)
     92         self.max_step = validate_max_step(max_step)
     93         self.rtol, self.atol = validate_tol(rtol, atol, self.n)
---> 94         self.f = self.fun(self.t, self.y)
     95         if first_step is None:
     96             self.h_abs = select_initial_step(

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/base.py in fun(t, y)
    136         def fun(t, y):
    137             self.nfev += 1
--> 138             return self.fun_single(t, y)
    139 
    140         self.fun = fun

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/base.py in fun_wrapped(t, y)
     18 
     19     def fun_wrapped(t, y):
---> 20         return np.asarray(fun(t, y), dtype=dtype)
     21 
     22     return fun_wrapped, y0

~/opt/anaconda3/lib/python3.9/site-packages/scipy/integrate/_ivp/ivp.py in <lambda>(t, x, fun)
    512         # additional parameters.  Pass in the original fun as a keyword
    513         # argument to keep it in the scope of the lambda.
--> 514         fun = lambda t, x, fun=fun: fun(t, x, *args)
    515         jac = options.get('jac')
    516         if callable(jac):

<lambdifygenerated-4> in _lambdifygenerated(_Dummy_26, t, kf, kb)
      1 def _lambdifygenerated(_Dummy_26, t, kf, kb):
----> 2     [y0, y1, y2] = _Dummy_26
      3     return [2*kb*y2**2 - 2*kf*y0**2*y1, kb*y2**2 - kf*y0**2*y1, -2*kb*y2**2 + 2*kf*y0**2*y1]

TypeError: cannot unpack non-iterable float object
```

moorepants replied:
This runs (but double check I didn't change the math).

solve_ivp takes t, y not y, t as args and the time is given as a tuple of (t0, tf) not an array. They changed several things about the api compared to odeint, which is a bit confusing.

```python
import sympy as sym
sym.init_printing()

def rhs(t, y, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, k = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(None, y, *k)

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((t, y) + k, ydot)  # EXERCISE: (y, t) + k

import numpy as np
from scipy.integrate import solve_ivp
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, (0, 10), y0, args=k_vals)
```

moorepants replied:
This removes the k values unpacking, maybe a bit clearer for learners.

```python
import sympy as sym
sym.init_printing()

def rhs(t, y, kf, kb):
    rf = kf * y[0]**2 * y[1]
    rb = kb * y[2]**2
    return [2*(rb - rf), rb - rf, 2*(rf - rb)]

y, (kf, kb) = sym.symbols('y:3'), sym.symbols('kf kb')
ydot = rhs(None, y, kf, kb)  # show that ydot is symbolic here

t = sym.symbols('t')  # not used in this case.
f = sym.lambdify((t, y, kf, kb), ydot)

import numpy as np
from scipy.integrate import solve_ivp
k_vals = np.array([0.42, 0.17])  # arbitrary in this case
y0 = [1, 1, 0]

solve_ivp(f, (0, 10), y0, args=k_vals)
```

bertiewooster replied:
Great, thanks!

bertiewooster committed:
Add lambdify + SciPy solve_ivp example

bertiewooster commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,391 @@
+      0.52048744 0.55430442 0.56645435 0.56979639 0.57044818 0.5705532 ]], 'y_events': None}
+```
+
+`ydot` is the derivative of the function `y`, and the value of `ydot` is given
+by the function `rhs` (right-hand side) for input values `y`, `kf`, and `kb`. We
+use {func}`~.lambdify` to convert the SymPy symbolic expression for `ydot` into
+a form that SciPy can evaluate numerically, `f`. Finally, we call SciPy's
+`solve_ivp` by passing it the function `f`, the interval of integration, the
+initial state, and the arguments to pass to the function `f`.
````
@moorepants I thought this code example merited some explanation. Could you please check that what I wrote explains it well?

bertiewooster committed:
Explain `solve_ivp` output

bertiewooster commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,393 @@
+use {func}`~.lambdify` to convert the SymPy symbolic expression for `ydot` into
+a form that SciPy can evaluate numerically, `f`. Finally, we call SciPy's
+`solve_ivp` by passing it the function `f`, the interval of integration, the
+initial state, and the arguments to pass to the function `f`. SciPy's
+`solve_ivp` returns a result containing `t` (time) points and corresponding `y`
+(numerical function result) values for each initial point in `y0`.
```
I believe I'm interpreting the [SciPy solve_ivp](https://docs.scipy.org/doc/scipy/reference/generated/scipy.integrate.solve_ivp.html#scipy.integrate.solve_ivp) documentation correctly that each sub-list in `y` corresponds to an initial point in `y0`. Then why are the results different for the first two sub-lists when each corresponds to a `y0` value of `1`?

moorepants replied:
I would interweave these explanations with each line of the code, i.e. state what each line is doing as you do it.

bertiewooster committed:
Add to blacklist (don't test) solve-ode.md if numpy or SciPy not installed

bertiewooster commented on sympy/testing/runtests.py:
```diff
@@ -729,6 +729,18 @@ def _get_doctest_blacklist():
             "sympy/parsing/sym_expr.py",
         ])
 
+    if import_module("scipy") is None:
+        # throws ModuleNotFoundError when scipy not installed
+        blacklist.extend(
+            ["doc/src/guides/solving/solve-ode.md",]
+        )
+
+    if import_module("numpy") is None:
+        # throws ModuleNotFoundError when scipy not installed
+        blacklist.extend(
+            ["doc/src/guides/solving/solve-ode.md",]
+        )
+
```
A doctest for solve-ode.md uses both SciPy and NumPy, so the doctest fails if those optional dependencies aren't installed. I thought it made more sense to add separate `if import_module("<module_name>") is None:` blocks for each (SciPy and NumPu) and add solve-ode.md to the blacklist in each block because
- there's another page which ([solve-numerically.md](https://github.com/sympy/sympy/pull/23853/files#diff-22796efbef97cd6b88b4d9cab6f86a61131f535414899b9f90d74baa6c1a3ce3) from #23853) which is added to the blacklist if scipy isn't installed, so 
- it seems more logical to me to have a block for each dependency rather than for each page.

I don't expect it will cause a problem for a page (solve-ode.md) to be added to the blacklist twice; please let me know if it would. An alternative is to have a single if block which added solve-ode.md to the blacklist if either dependency isn't installed.

bertiewooster committed:
Improve link text to "example from the field of chemical
kinetics"

moorepants commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,393 @@
+[SciPy's](https://docs.scipy.org/doc/scipy/index.html) fast numerical ODE
+solving is
+1. set up an ODE in SymPy
+2. convert it to a lambda function using {func}`~.lambdify`
```
```suggestion
2. convert it to a numerical function using {func}`~.lambdify`
```

moorepants commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,393 @@
+solving is
+1. set up an ODE in SymPy
+2. convert it to a lambda function using {func}`~.lambdify`
+3. solve it numerically using SciPy's `solve_ivp`.
```
```suggestion
3. solve the initial value problem by numerically integrating the ODE using SciPy's `solve_ivp`.
```

moorepants commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,393 @@
+```
+
+Here is an [example from the field of chemical
+kinetics](https://www.sympy.org/scipy-2017-codegen-tutorial/notebooks/25-chemical-kinetics-intro.html):
````
```suggestion
kinetics](https://www.sympy.org/scipy-2017-codegen-tutorial/notebooks/25-chemical-kinetics-intro.html) 
where the nonlinear ordinary differential equations take this form:

$$
r_f = & k_f y_0(t)^2  y_1(t) \\
r_b = & k_b y_2(t)^2 \\
\frac{d y_0(t)}{dt} = & 2(r_b - r_f) \\
\frac{d y_1(t)}{dt} = & r_b - r_f \\
\frac{d y_2(t)}{dt} = & 2(r_f - r_b)
$$

and

$$

\vec{y}(t) = \begin{bmatrix}
y_0(t) \\
y_1(t) \\
y_2(t)
\end{bmatrix}
$$
```

moorepants commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,393 @@
+...     rb = kb * y[2]**2
+...     return [2*(rb - rf), rb - rf, 2*(rf - rb)]
+>>> y, (kf, kb) = symbols('y:3'), symbols('kf kb')
+>>> ydot = rhs(None, y, kf, kb)
```
```suggestion
>>> ydot = rhs(None, y, kf, kb)
>>> ydot
```

Display ydot so that they see it is symbolic at this stage.

moorepants replied:
Also, passing in `None` for `t` is probably not clear for a learner. Using the same `rhs` to both construct the symbolic form and evaluate the numerics is a bit of a trick that will be lost on learners. I think creating the symbolic form separately and explicitly, as you do in prior examples, is more clear for a learner. If you do that, then you can avoid these tricks for shortening the code.

moorepants commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,393 @@
+>>> t = symbols('t') # not used in this case
+>>> f = lambdify((t, y, kf, kb), ydot)
+>>> k_vals = np.array([0.42, 0.17]) # arbitrary in this case
+>>> y0 = [1, 1, 0]
```
call this the initial condition somewhere

moorepants commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,393 @@
+      0.73975628 0.72284779 0.71677282 0.71510181 0.71477591 0.7147234 ]
+     [0.         0.00141031 0.01524469 0.13110138 0.31879762 0.44609392
+      0.52048744 0.55430442 0.56645435 0.56979639 0.57044818 0.5705532 ]], 'y_events': None}
+```
````
We have the matplolib extension installed so you can create graphs. This data is more easily understood if you make a simple plot. See the beam module docs for examples of using the plotting extension (or the plotting module).

bertiewooster replied:
Thanks. I made a graph which displays properly in a Jupyter Notebook. But I'm having trouble getting the plot to display properly via Sphinx. I was able to get a simple matplotlib example to display, but not the ODE solution. Also, may not be relevant once the ODE plot works, but if I make a second simple plot with different values, it shows the first plot again:

<img width="530" alt="matplotlib for ODE" src="https://user-images.githubusercontent.com/1335522/191617582-3707e842-1467-4769-b4b3-44bc862ffd6d.png">

moorepants replied:
There are flags to control whether figures are closed. We got it working in the beam pages.

bertiewooster replied:
One problem is that any plot code block containing scipy fails due to:

`ModuleNotFoundError: No module named 'scipy'`

(A simple plot block like `plt.plot([11, 12, 13], [16, 15, 14])` renders a graph image fine if I comment out the scipy import; numpy and matplotlib imports work fine.) Is there some way to make a package available inside a plot block? Perhaps this [configuration option](https://matplotlib.org/3.1.3/devel/plot_directive.html#configuration-options):

> plot_working_directory
>     By default, the working directory will be changed to the directory of the example, so the code can get at its data files, if any. Also its path will be added to [sys.path](https://docs.python.org/3/library/sys.html#sys.path) so it can import any helper modules sitting beside it. This configuration option can be used to specify a central directory (also added to [sys.path](https://docs.python.org/3/library/sys.html#sys.path)) where data files and helper modules for all code are located.

bertiewooster replied:
> There are flags to control whether figures are closed. We got it working in the beam pages.

Thanks--yes, by adding `:context: close-figs` to each plot directive as in the [beam module](https://github.com/sympy/sympy/blob/26f7bdbe3f860e7b4492e102edec2d6b429b5aaf/sympy/physics/continuum_mechanics/beam.py#L1314), different figures can be created. Probably not needed for this page because I plan on including only one plot; good to know how to do.

bertiewooster replied:
Turned out that SciPy needed to be a dependency for the docs, by adding `scipy` to `doc/requirements.txt`.

bertiewooster committed:
lambad -> numerical function

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

bertiewooster committed:
clarify what solve_ivp does

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

bertiewooster committed:
write ODE in math notation

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

bertiewooster committed:
Display ydot so that visitors know it is symbolic at this stage

Co-authored-by: Jason K. Moore <moorepants@gmail.com>

bertiewooster committed:
Make several suggested improvements for SciPy example

bertiewooster committed:
Merge branch 'master' into solve-ode

bertiewooster committed:
Trying to fix error
SyntaxError: list indices must be integers or slices, not tuple; perhaps you missed a comma?

bertiewooster committed:
Merge branch 'solve-ode' of github.com:bertiewooster/sympy into solve-ode

bertiewooster committed:
Fix formatting of runtests.py

bertiewooster committed:
Add plot of chemical kinetics results

bertiewooster committed:
Trying to get plot to show

bertiewooster committed:
Moved entire example into .. plot:: block

asmeurer commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,467 @@
+```
+
+```{eval-rst}
+.. plot:: 
````
It should be possible to do this directly with myst. The rst directives translate directly into markdown syntax. See https://myst-parser.readthedocs.io/en/v0.13.6/using/syntax.html#directives-a-block-level-extension-point

bertiewooster replied:
Yeah, I tried, using that format:

````
```{plot}
---
format: doctest 
include-source: True
---
>>> plt.plot([1, 2, 3], [4, 5, 6]) # doctest: +SKIP
```
````

and a slightly different one:

````
```{plot}
:format: doctest 
:include-source: True

>>> plt.plot([1, 2, 3], [4, 5, 6]) # doctest: +SKIP
```
````

and neither gave any output.

I asked on [Write the Docs (Slack for technical writing)](https://www.writethedocs.org/slack/) and the Executable Books lead developer @chrisjsewell replied:

> Yeh so from here: https://github.com/matplotlib/matplotlib/blob/269c0b94b4fcf8b1135011c1556eac29dc09de15/lib/matplotlib/sphinxext/plot_directive.py#L359
> the default `TEMPLATE` is hard-coded to be in rST.
> It looks like you can override this with the `plot_template` configuration option,
> or obviously you can raise a feature request with them to include such a template :sweat_smile:

Which I believe means someone would have to create a `TEMPLATE` which works for Markdown.

asmeurer replied:
I actually remember having an issue with plotting in my custom functions guide too. That's good to know. So for now if we want to include plots here, we need to rewrite this page to use RST?

bertiewooster replied:
No, thankfully it's fine for the page to be in Markdown as long as the plot directive is in RST--[a simple plot block renders a graph image fine](https://output.circle-artifacts.com/output/job/03cc4694-a200-4587-aa2b-bdd4fe6dd463/artifacts/0/doc/_build/html/guides/solving/solve-ode.html#numerically-solve-an-ode-in-scipy) using the wrapper:

````
```{eval-rst}
```
````

I think the remaining critical problem is importing scipy within the plot directive.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,467 @@
+  >>> solution = scipy.integrate.solve_ivp(f, (0, 10), y0, t_eval=t_eval, args=k_vals) # call SciPy's solve_ivp by passing it the function f, the interval of integration, the initial state, and the arguments to pass to the function f
+  >>> solution
+      {'message': 'The solver successfully reached the end of the integration interval.', 'nfev': 68, 'njev': 0, 'nlu': 0, 'sol': None, 'status': 0, 'success': True, 't': [ 0.          0.20408163  0.40816327  0.6122449   0.81632653  1.02040816
+        1.2244898   1.42857143  1.63265306  1.83673469  2.04081633  2.24489796
```
You can replace unimportant chunks of the output with `...`

bertiewooster replied:
Thanks. I plan on taking that one step further: Assuming the plot will work, I don't think it's even necessary to include the solution output at all.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,467 @@
+  >>> k_vals = np.array([0.42, 0.17]) # arbitrary in this case
+  >>> y0 = [1, 1, 0] # initial condition (initial values)
+  >>> t_eval = np.linspace(0, 10, 50) # evaluate integral from t = 0-10 for 50 points
+  >>> solution = scipy.integrate.solve_ivp(f, (0, 10), y0, t_eval=t_eval, args=k_vals) # call SciPy's solve_ivp by passing it the function f, the interval of integration, the initial state, and the arguments to pass to the function f
```
It's easier to read long comments like this in doctest blocks when they are on their own line before the code, since text in doctest blocks does not text wrap, so any line that is too long will require you to scroll right to read it.

bertiewooster replied:
Great, thanks for the tip. I thought the doctest would complain if you inserted a line with just a comment; seems fine (at least running doctest locally) here.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,467 @@
+3. solve the initial value problem by numerically integrating the ODE using
+   SciPy's `solve_ivp`.
+
+```{warning}
````
I don't know if this needs to be repeated here.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,467 @@
+7*sin(3*x) + pi*cos(3*x)
+```
+
+### Numerically Solve an ODE in SciPy
````
I think this should be a level 2 header. It's not related to "use the solution result".

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,467 @@
+
+Unlike other solving functions, {func}`~.dsolve` returns an {class}`~.Equality`
+(equation) formatted as, for example, `Eq(y(x), C1*sin(3*x) + C2*cos(3*x))`
+which is equivalent to the mathematical notation $y(x) = C1 \sin(3x) + C2
```
Make sure to subscript the constants in LaTeX (`C_1`). You can also generally just use `print(latex(expr))` on an expression to get a LaTeX form to put in the docs.

bertiewooster committed:
Plot generated as long as scipy not used

bertiewooster committed:
Clarify symbol creation

bertiewooster committed:
Add scipy to doc/requirements.txt so can be imported for figure
Add annotations to figure

bertiewooster committed:
Rephrase explanation of what solve_ivp returns

bertiewooster committed:
Put backticks around t_eval

bertiewooster committed:
Add link from ODE solver module to guide

bertiewooster committed:
Merge branch 'master' into solve-ode

asmeurer commented on doc/src/guides/solving/index.md:
```diff
@@ -15,7 +15,8 @@ Learn how to use SymPy computer algebra system to:
 | [ Solve a system of equations (linear or nonlinear) algebraically ](solve-system-of-equations-algebraically.md)              | $x^2 + y = 2z, y = -4z$ | $\{(x = -\sqrt{6z}, y = -4z),$ ${(x = \sqrt{6z}, y = -4z)\}}$                                                                                        |
 |  [Solve one or a system of equations numerically](solve-numerically.md)                           | $\cos(x) = x $ | $ x \approx 0.739085133215161$                                                                                           |
-|  {func}`Solve an ordinary differential equation algebraically <sympy.solvers.ode.dsolve>`   | $y''(x) + 9y(x)=0 $ | $ y(x)=C_{1} \sin(3x)+ C_{2} \cos(3x)$                                                    |
+|  [Solve an ordinary differential equation algebraically](solve-ode.md)   | $y''(x) + 9y(x)=0 $ | $ y(x)=C_{1} \sin(3x)+ C_{2} \cos(3x)$                                                    |
```
This appears to have broken the rendering of the table.

asmeurer commented on doc/src/guides/solving/index.md:
```diff
@@ -15,7 +15,8 @@ Learn how to use SymPy computer algebra system to:
 |  [Solve one or a system of equations numerically](solve-numerically.md)                           | $\cos(x) = x $ | $ x \approx 0.739085133215161$                                                                                           |
-|  {func}`Solve an ordinary differential equation algebraically <sympy.solvers.ode.dsolve>`   | $y''(x) + 9y(x)=0 $ | $ y(x)=C_{1} \sin(3x)+ C_{2} \cos(3x)$                                                    |
+|  [Solve an ordinary differential equation algebraically](solve-ode.md)   | $y''(x) + 9y(x)=0 $ | $ y(x)=C_{1} \sin(3x)+ C_{2} \cos(3x)$                                                    |
+
```
I think the problem is this extra newline here.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+
+## Solve an Ordinary Differential Equation (ODE)
+
+Here is an example of solving an ordinary differential equation algebraically
```
```suggestion
Here is an example of solving the above ordinary differential equation algebraically
```

asmeurer commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,423 @@
+
+```py
+>>> from sympy import checkodesol
+>>> solution = result.rhs
````
It's not necessary to do this. `checkodesol` can accept the direct Eq output of `dsolve`.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+tells whether substituting the solution into the ODE results in `0`, indicating
+the solution is correct.
+
+## Guidance
```
I'm a little unclear how the "Guidance" section is supposed to work on these pages. How come on this page the subsequent sections are on the same level as "Guidance", but on the "solving systems of equations numerically" page several sections are under "Guidance"?

bertiewooster replied:
The intent of the "Guidance" section is to have zero or more subsections that provide tips on how to set up using the function in question. 

- On this ODE page, I identified only one tip, so there's just one subsection, "Input Format"; so "Guidance" actually has one child section. The subsequent sections are not (in my way of thinking) that kind of guidance, so they're at the same level (`h2`) as "Guidance".
- On the ["solving systems of equations numerically" page](https://docs.sympy.org/dev/guides/solving/solve-numerically.html), I identified several such tips, so they're all children sections of "Guidance".

The (`h2`) sections after "Guidance" are typically "Use the Solution Result", "Tradeoffs", etc. I admit it's a somewhat subjective decision what counts as Guidance. I used a template with sections and slotted content into what I thought was the most logical section. We can revisit if needed.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+You can define the function to be solved for in two ways. The subsequent syntax
+for specifying initial conditions depends on your choice.
+
+## Define a Function Without Including Its Independent Variable
```
I think this is supposed to be level 4 (and the same with the next several headers).

bertiewooster replied:
My intention was to have

1. Define a Function Without Including Its Independent Variable
2. Define a Function of an Independent Variable

as level 2, with the idea that each is an _option_ (as you suggest below) for an entire approach to using dsolve--setting up `dsolve`, adding initial conditions or boundary conditions, and (for option 2) a caveat. The subsequent sections (Use the Solution Result, etc.) apply to either option 2.

If you think it would be helpful, I could create a parent level 2 and put the two options below it as level 3:
## Options to Define an ODE
### Option 1: Define a Function Without Including Its Independent Variable
### Option 2: Define a Function of an Independent Variable

bertiewooster replied:
Move to a new h2 Options to Define an ODE
You can define the function to be solved for in two ways. The subsequent syntax for specifying initial conditions depends on your choice.

Could move options 1 and 2 under Guidance instead

Rename "Input Format" to Defining Derivatives

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+
+## Define a Function Without Including Its Independent Variable
+
+As in the example above, you can define a function without including its
```
It's a little confusing what this means. It took me a moment to realize this is talking specifically about the line `f = Function('f')`. The independent variable (x) is used elsewhere, just not on that line.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+If your differential equation(s) have initial or boundary conditions, specify
+them with the {func}`~.dsolve` optional argument `ics`. It should be given in
+the form of `{f(x0): y0, f(x).diff(x).subs(x, x1): y1}` and so on where, for
+example, the value of `f(x)` at $x = $ `x0` is `y0`. For power series solutions,
```
Don't mix LaTeX and code like this. Use either one or the other.

asmeurer replied:
Should also note here that `f'(x1) = y1` (that's the one that is less clear).

bertiewooster replied:
My interpretation was that

- `code` is for code, e.g. `dsolve`, as in the [documentation style guide](https://docs.sympy.org/dev/contributing/documentation-style-guide.html#code)
- LaTeX is for math, e.g. $\pi$, similar to the [docstring style guide](https://docs.sympy.org/dev/contributing/docstring.html#docstrings-for-classes-that-are-mathematical-functions)

Does that make sense?

bertiewooster replied:
For e.g. x = x0, make 0 a subscript. Use math.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,423 @@
+```
+
+Here is an example of setting the initial value for the derivative of a
+function, namely $f'(1) = 2$:
````
Even though it's supported, I don't think a typical use-case would be to underspecify the initial conditions like this. I would either also add a second condition or change this example to use a single ODE.

bertiewooster replied:
Either specify an initial condition for g'(x), or just use one equation.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+Note that you supply the functions to be solved for as a list as the second
+argument of {func}`~.dsolve`, here `[f(x), g(x)]`.
+
+### Specify Initial (Boundary) Conditions
```
```suggestion
### Specify Initial Conditions or Boundary Conditions
```

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+
+### Specify Initial (Boundary) Conditions
+
+If your differential equation(s) have initial or boundary conditions, specify
```
It should be stressed here that both initial conditions and boundary conditions are treated in the same way (even though the argument is called `ics`).

asmeurer commented on doc/src/guides/solving/solve-ode.md:
````diff
@@ -0,0 +1,423 @@
+`x(t)(t)`:
+
+```py
+>>> dsolve(x(t).diff(), x)
````
The no-argument variant of `diff()` that doesn't specify the symbol to differentiate wrt is provided as a convenience, but it's generally considered bad practice, so we should avoid using it in documentation.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+### Extract the Result for One Solution and Function
+
+You can extract the result from an {class}`~.Equality` using the right-hand side
+property {any}`rhs <sympy.core.relational.Relational.rhs>`:
```
I don't know if we should caution here that the result may not actually be solved for f(x). For cases when SymPy cannot solve the equation explicitly, the solution will be returned in implicit form.

bertiewooster replied:
Plan: Add a section and example where SymPy can't find explicit solution--just an implicit solution. Just find a function that can't be solved, make into a differential equation for y; for example, x - cos(x) which can't be solved algebraically.

asmeurer replied:
Here's an example 

```
>>> dsolve(f(x).diff(x) + exp(-f(x))*f(x))
Eq(Ei(f(x)), C1 - x)
```

Here Ei is the exponential integral function, which doesn't have a closed-form inverse as far as I know. It looks like dsolve doesn't do anything in this case. It just returns a solution that's something equals something, where either side could contain f(x) (i.e., an [implicit solution](https://en.wikipedia.org/wiki/Implicit_function)).

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+### Extract the Result for Multiple Function-Solution Pairs
+
+If you are solving a system of equations with multiple unknown functions,
+{func}`~.dsolve` will return a nested list of equalities, the outer list
```
It only returns a nested list when there are multiple different solution sets. When there is only one solution set it just returns a single list (for example, if you replace $y'(x)^2 = z(x)^2$ with $y'(x) = z(x)^2$  in this example).

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+
+## Ordinary Differential Equation Solving Hints
+
+### Return Unevaluated Integrals
```
We really just need to add this as a flag to dsolve so you don't have to mess with hints to do this.

bertiewooster replied:
That would be nice.

asmeurer commented on doc/src/guides/solving/solve-ode.md:
```diff
@@ -0,0 +1,423 @@
+ypp = diff(y(x), x, 2)
+ypp = Derivative(y(x), x, x)
+ypp = Derivative(y(x), x, 2)
+ypp = Derivative(Derivative(y(y(x), x), x)
```
```suggestion
ypp = Derivative(Derivative(y(x), x), x)
```

bertiewooster committed:
Address about half of review comments

bertiewooster committed:
Change page structure
Add implicit solution example

bertiewooster committed:
Add example for: only one solution set to a system of equations with multiple unknown
functions

bertiewooster committed:
Use only one function and equation for setting initial value of derivative

bertiewooster committed:
Remove reference to example above

bertiewooster committed:
Never use x as independent variable

bertiewooster committed:
Merge branch 'master' into solve-ode

bertiewooster committed:
Remove unintended label solving-guide-inequalities

bertiewooster committed:
Fix link from ODE module to solving guide

bertiewooster committed:
Merge branch 'master' into solve-ode

asmeurer commented:
There seem to be a lot of changes to runtests.py here? Did you run some sort of autoformatter on it?

asmeurer commented on doc/src/conf.py:
```diff
@@ -396,6 +396,7 @@
 # Enable links to other packages
 intersphinx_mapping = {
     'matplotlib': ('https://matplotlib.org/stable/', None),
+    'python': ('https://docs.python.org', None),
```
Where is there a cross-reference to the Python docs?

bertiewooster replied:
There used to be a link to a Python function. I removed the link, so I just removed the cross-reference.

bertiewooster commented:
> There seem to be a lot of changes to runtests.py here? Did you run some sort of autoformatter on it?

Oops, yes. I'll undo the autoformatter changes, and just make the necessary changes.

bertiewooster committed:
Remove intersphinx link to Python

bertiewooster committed:
runtests.py: Undo auto-formatting, re-add changes to blacklist for scipy, numpy

bertiewooster committed:
Fix syntax error in runtests.py
Rephrase a sentence in ODE page

bertiewooster committed:
Fix comment about which module
Make square bracket placement consistent

asmeurer commented on doc/requirements.txt:
```diff
@@ -4,6 +4,7 @@ matplotlib
 matplotlib-inline
 mpmath
 myst-parser
+scipy
```
Do we actually need scipy to build the docs? If not, we shouldn't put it here.

bertiewooster replied:
As we discussed, yes we do, because we use scipy to create the plot.

bertiewooster committed:
Merge branch 'master' into solve-ode
